### DISCUSSION TITLE
How to get a provider for web3.js?

### DISCUSSION THREAD

**Discussion by orangeqq** [WalletConnect/walletconnect-monorepo#2963] [Developer Support]
In version 1.0, I used web3Modal.connect() to obtain a provider for web3.js as follows:
```javascript
web3Modal = new Web3Modal({
    cacheProvider: false, // optional
    providerOptions, // required
    disableInjectedProvider: false, // optional. For MetaMask / Brave / Opera.
});
provider = await web3Modal.connect();
const web3 = new Web3(provider);
const accounts = await web3.eth.getAccounts();
``` 
Now I'm trying to upgrading to v2.0. with Wagmi like this:
[https://docs.walletconnect.com/2.0/web/web3modal/html/wagmi/installation#implementation](https://docs.walletconnect.com/2.0/web/web3modal/html/wagmi/installation#implementation)

At last line of the implementation part above, it seems I got a web3modal.
```javascript
const web3modal = new Web3Modal({ projectId }, ethereumClient)
``` 
But when I'm trying to get a provider through it like what I did in v1.0
```javascript
provider = await web3Modal.connect();
``` 
I got error in browser console:
```javascript
TypeError: web3Modal.connect is not a function. 
``` 

How can I get a provider in Wagmi and give it to web3.js?

**Comment by boidushya** [comment]
Hey! Please refer to [this thread](https://github.com/orgs/WalletConnect/discussions/2920)

**Comment by orangeqq** [comment]
Thank you for your reply, sir. @boidushya 
I think it is a perfect way to get a provider.
But things become a little complicated here. I'm using Vanilla JS and I'm not yet know how to use EthereumProvider in it.
This is [my another thread](https://github.com/orgs/WalletConnect/discussions/2939) about this.

So I'm trying another way with wagmi.
[Wagmi docs](https://wagmi.sh/core/connectors/walletConnect) shows I can get [WalletConnect Ethereum Provider] by using WalletConnectConnector, but I have no idea how to get it.

<img width="849" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/4606809/2801de40-e55e-4f0f-bbbb-88a881e4a3b0">

**Reply by boidushya** [reply]
Hey you can use `WalletConnectConnector` as mentioned under usage [here](https://wagmi.sh/react/connectors/walletConnect#usage)

**Reply by orangeqq** [reply]
Yes, I'm using the WalletConnectConnector, but the usage part doesn't mention the WalletConnect Ethereum Provider.

I created a discussion [there](https://github.com/wagmi-dev/wagmi/discussions/2685)

**Reply by boidushya** [reply]
Sounds good! Wagmi folks can help you out with that

**Comment by orangeqq** [comment]
I'm modifying this sample to get a provider for web3.js
https://github.com/WalletConnect/web3modal-examples/blob/main/web3modal-wagmi-html-cdn/main.js

The sample works well. But obviously, the [web3Modal] hasn't a connect() function.
How to use [WalletConnect Ethereum Provider] in this sample?

**Reply by boidushya** [reply]
Check [this](https://docs.walletconnect.com/2.0/web/web3modal/html/wagmi/installation) out for using web3modal with HTML

**Reply by SaadNiaziDev** [reply]
first u need to pop-up web3modal after scanning or connecting via metamask will provide u provider. You should use openModal function of web3modal
Go through above docs link provided by @boidushya

**Comment by orangeqq** [comment]
Thank you everyone, my problem has been resolved!
In the end, I decided not to pass the provider to web3.js. I discovered that the session obtained through [WalletConnectModal (with Sign API)](https://docs.walletconnect.com/2.0/web/walletConnectModal/sign/html/actions#connect), can be directly used to [send JSON-RPC requests](https://docs.walletconnect.com/2.0/web/walletConnectModal/sign/html/actions#request). And that's exactly what I needed.
I hope this information can help others who have the same problem as me.

**Reply by dowhileitworks** [reply]
Hi can you post the full code now? i am facing the same issue
tnks

**Reply by orangeqq** [reply]
Sure!

1. add the js file as a module in html
```javascript
<script src="../transfermobile.js" type="module"></script>
```

2. in the js file, create the web3Modal object
```javascript
import { WalletConnectModalSign } from "https://unpkg.com/@walletconnect/modal-sign-html@2.5.8";

export function walletConnectInitV2() {
    web3Modal = new WalletConnectModalSign({
        projectId: "xxx",
        metadata: {
            name: "xxx",
            description: "xxx",
            url: "https://xxx/",
            icons: ["https://xxx/favicon.ico"],
        }
    });
}
```
3. create session
```javascript
async function createMobileSession() {
    try {
        session = await web3Modal.connect({
            requiredNamespaces: {
                eip155: {
                    methods: ["eth_sendTransaction", "personal_sign"],
                    chains: ["eip155:1"],
                    events: ["chainChanged", "accountsChanged", "connect", "disconnect"],
                },
            },
        });
    } catch (err) {
        console.error(err);
        return;
    } finally {
        return session;
    }
}
```

4. send the jsonrpc request
```javascript
async function mobileTransferV2(toAccount, contractAddress, amount, decimal) {
    session = await createMobileSession();
    fromAccount = session.namespaces.eip155.accounts[0].split(':')[2].toLowerCase();
    let tokenAmmout = "0x" + (amount * (10**decimal)).toString(16);
    let transactionParam = {
        from: fromAccount,
        to: contractAddress,
        data: getDataFieldValue(toAccount, tokenAmmout),
    };
    try {
        const response = await web3Modal.request({
            topic: session.topic,
            chainId: 'eip155:1',
            request: {
                id: 1,
                jsonrpc: '2.0',
                method: 'eth_sendTransaction',
                params:[transactionParam]
            }
        });
    } catch (err) {
        console.error(err);
    }
}
```

5. used function which based on web3.js
```javascript
function getDataFieldValue(tokenRecipientAddress, tokenAmount) {
    const web3 = new Web3();
    const TRANSFER_FUNCTION_ABI = {"inputs":[{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"amount","type":"uint256"}],"name":"transfer","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"nonpayable","type":"function"};
    return web3.eth.abi.encodeFunctionCall(TRANSFER_FUNCTION_ABI, [
        tokenRecipientAddress,
        tokenAmount
    ]);
}
```

That's all.

**Reply by wavywtf** [reply]
this is your setup for a wallet dap or just a regular dapp?

**Reply by orangeqq** [reply]
It's a regular dapp.
